raft/raft.go

```diff
@@ -49,36 +49,86 @@ func (st StateType) String() string {
 	return stmap[uint64(st)]
 }
 
+const (
+	ProgressStateProbe ProgressStateType = iota
+	ProgressStateReplicate
+	ProgressStateSnapshot
+)
+
+type ProgressStateType uint64
+
+var prstmap = [...]string{
+	"ProgressStateProbe",
+	"ProgressStateReplicate",
+	"ProgressStateSnapshot",
+}
+
+func (st ProgressStateType) String() string { return prstmap[uint64(st)] }
+
 type Progress struct {
 	Match, Next uint64
-	Wait        int
-	// If the last sent to the Progress failed and reported
-	// by the link layer via MsgUnreachable, Unreachable will be set.
-	// If the Progress is unreachable, snapshot and optimistically append
-	// will be disabled.
-	// Unreachable will be unset if raft starts to receive message (msgAppResp,
-	// msgHeartbeatResp) from the remote peer of the Progress.
-	Unreachable bool
+	// When in ProgressStateProbe, leader sends at most one replication message
+	// per heartbeat interval. It also probes actual progress of the follower.
+	//
+	// When in ProgressStateReplicate, leader optimistically increases next
+	// to the latest entry sent after sending replication message. This is
+	// an optimized state for fast replicating log entries to the follower.
+	//
+	// When in ProgressStateSnapshot, leader should have sent out snapshot
+	// before and stops sending any replication message.
+	State ProgressStateType
+	// Paused is used in ProgressStateProbe.
+	// When Paused is true, progress stops sending replication message.
```

Comment:
When Paused is true, raft should pause sending replication message to this peer.